### PR TITLE
Add Ability to Specify All Public Keys When Exiting Validators

### LINKED ISSUE #8383: Add Ability to Exit All Validators
# 🚀 Feature Request

### Description

Currently, stakers with many validators cannot perform mass voluntary exits as the CLI command is interactive

### Describe the solution you'd like

Allow performing automated exits by skipping the confirmation prompts during the CLI command for validators.

### Describe alternatives you've considered

Someone advanced enough to use the various flags to skip confirmations knows what they are doing, and there is no reason we should disallow automatic exits.

### PR DESCRIPTION
**Pull request by rauljordan**
**What does this PR do? Why is it needed?**

Currently, validators cannot exit all accounts if desired. This PR adds a flag `--exit-all` to the validator client. The staker still needs to confirm a prompt given the dangerous nature of the action.

**Which issues(s) does this PR fix?**

Fixes #8383

### CODE REVIEW

**Review by terencechain — DISMISSED** [review]

**Review by rkapka — COMMENTED** [review]
From the issue:

> So basically one command to request exit for all validators at the same time. Maximum a yes/no question but perhaps not even that. The ones who looks up and executes this command surely knows he is about to exit all validators wether its 1 x 32ETH or 500 x 32ETH instantly.

I think what this means is an option to exit all validators without even having to select anything. Just type the command, maybe display a yes/no confirmation and exit everything. Do we want to support that? It's risky.

**Review by rkapka — DISMISSED** [review]
Blocking the PR for now as I don't think this fixes the issue, as this is not what OP wants.

**Comment by makultar** [comment]
> Comment from the issue's creator @makultar:
> 
> > So basically one command to request exit for all validators at the same time. Maximum a yes/no question but perhaps not even that. The ones who looks up and executes this command surely knows he is about to exit all validators wether its 1 x 32ETH or 500 x 32ETH instantly.
> 
> I think what this means is an option to exit all validators without even having to select anything. Just type the command, maybe display a yes/no confirmation and exit everything. Do we want to support that? It's risky.

> Blocking the PR for now as I don't think this fixes the issue, as this is not what OP wants.
Holddddddd your horses! 

I think something like _--skip-voluntary-exit-confirmation_ is exactly was I am looking for. Or at least on the right path. 

Hmm I dont agree with "Its risky". Staking itself is already risky, i.e depositing 50 000 USD+, making sure the system works flawlessly, that you dont get hacked, making sure your mnemonic is with you x years while withdrawal is disabled etc etc. There is just tons of various risks. Making an exit to keep your funds safe surely is not. Then lets make sure its well described and with several warnings around that description?

No one is dumb enough to execute a command thats wrapped around
*** WARNING * WARNING * WARNING * WARNING * WARNING * WARNING *** 
*** THE FOLLOWING COMMAND WILL EXIT ALL YOUR VALIDATORS AT ONCE ***
*** WARNING * WARNING * WARNING * WARNING * WARNING * WARNING *** 
There is no real risk. If you read this WARNING, and still decide to go ahead with a: YES -- thats on you. We are adult, not kids.

**Comment by makultar** [comment]
> Comment from the issue's creator @makultar:
> 
> > So basically one command to request exit for all validators at the same time. Maximum a yes/no question but perhaps not even that. The ones who looks up and executes this command surely knows he is about to exit all validators wether its 1 x 32ETH or 500 x 32ETH instantly.
> 
> I think what this means is an option to exit all validators without even having to select anything. Just type the command, maybe display a yes/no confirmation and exit everything. Do we want to support that? It's risky.

> Blocking the PR for now as I don't think this fixes the issue, as this is not what OP wants.

Holddddddd your horses! 

I think something like _--skip-voluntary-exit-confirmation_ is exactly was I am looking for. Or at least on the right path. 

Hmm I dont agree with "Its risky". Staking itself is already risky, i.e depositing 50 000 USD+, making sure the system works flawlessly, that you dont get hacked, making sure your mnemonic is with you x years while withdrawal is disabled etc etc. There is just tons of various risks. Making an exit to keep your funds safe surely is not. Very much the opposite. Then lets make sure its well described and with several warnings around that description so no one else can be blamed?

No one is dumb enough to execute a command thats wrapped around

*** **WARNING * WARNING * WARNING * WARNING * WARNING * WARNING** *** 
_*       THE FOLLOWING COMMAND WILL EXIT ALL YOUR VALIDATORS AT ONCE        *_
*** **WARNING * WARNING * WARNING * WARNING * WARNING * WARNING*** **

There is no real risk. If you read this WARNING x 12, and still decide to go ahead with a: YES -- thats on you. We are adults, not kids. Who is there to protect? I doubt there is kids or adults with severe brain damage staking 50k that dont know what they are doing :P

Edit: Typos hehe

**Comment by rkapka** [comment]
Copying from Discord:

IIRC, right now you have to manually enter all keys that you want to exit. accounts delete has this nice dropdown that you can choose from, with an Select all option. Maybe we can utilize that, combined with the flag that you introduced?

**Comment by makultar** [comment]
> Copying from Discord:
> 
> IIRC, right now you have to manually enter all keys that you want to exit. `accounts delete` has this nice dropdown that you can choose from, with an `Select all` option. Maybe we can utilize that, combined with the flag that you introduced?

I would be very happy if the following command was possible
`/usr/local/bin/validator accounts voluntary-exit --wallet-dir=/var/lib/prysm/validator _--select-all --skip-voluntary-exit-confirmation_`

**Letting you only confirm with a yes/no:**
*** **WARNING * WARNING * WARNING * WARNING * WARNING * WARNING** *** 
_*       THE FOLLOWING COMMAND WILL EXIT ALL YOUR VALIDATORS AT ONCE        *_
_*                                            ARE YOU SURE? Yes/No                                                  *_
*** **WARNING * WARNING * WARNING * WARNING * WARNING * WARNING*** **

=> Exits

**Comment by makultar** [comment]
For someone thats not a techie, like my family, wife, or future kids. I dont want this to be a complicated task or menu. My instructions to them will be very simply.

1. Login into server 127.0..1
2. Copypaste and press enter: 
3. /usr/local/bin/validator accounts voluntary-exit --wallet-dir=/var/lib/prysm/validator _ --select-all --skip-voluntary-exit-confirmation _
4. Type "Yes" => Enter

**Comment by makultar** [comment]
For someone thats not a techie, like my family, wife, or future kids. I dont want this to be a complicated task or menu. My instructions to them will be very simply.

1. Login into server 127.0..1
2. Copypaste and press enter: 
/usr/local/bin/validator accounts voluntary-exit --wallet-dir=/var/lib/prysm/validator **--select-all --skip-voluntary-exit-confirmation**
3. Type "Yes" when asked to confirm => Enter

Not a long restaurant menu going through all kinds of options, opening up for real real problems.

**Comment by rauljordan** [comment]
Thanks @makultar totally understand your rationale now. All you want is a simple command that allows you specify you want to exit all public keys via a flag, then a simple confirmation that looks scary, and then you're done. We agree this is an important feature to safeguard your eth in case something happens to you and a family member needs to perform the action, thank you for bringing it to our attention, we will amend to do as you requested

**Comment by makultar** [comment]
> Thanks @makultar totally understand your rationale now. All you want is a simple command that allows you specify you want to exit all public keys via a flag, then a simple confirmation that looks scary, and then you're done. We agree this is an important feature to safeguard your eth in case something happens to you and a family member needs to perform the action, thank you for bringing it to our attention, we will amend to do as you requested

Wow thats great! Thanks a ton for this. Cool and kind of you guys. I will launch a few testnet validators and test this feature for sure.

File: validator/accounts/accounts_exit.go [hunk 1]
```diff
@@ -99,54 +99,64 @@ func interact(
...
 			}
 		}
+	} else {
+		rawPubKeys = make([][]byte, len(validatingPublicKeys))
+		formattedPubKeys = make([]string, len(validatingPublicKeys))
+		for i, pk := range validatingPublicKeys {
+			rawPubKeys[i] = pk[:]
+			formattedPubKeys[i] = fmt.Sprintf("%#x", bytesutil.Trunc(pk[:]))
+		}
+		fmt.Printf("About to perform a voluntary exit of %d accounts\n", len(rawPubKeys))
```
Comment (rkapka) [inline validator/accounts/accounts_exit.go:159]: This needs a test.

**Review by rkapka — CHANGES_REQUESTED** [review]
Please test the new functionality.

**Comment by rauljordan** [comment]
Done @rkapka

**Comment by rauljordan** [comment]
Ready again @rkapka

**Review by rkapka — APPROVED** [review]
Looks great, thanks!

**Pull request merged**